### PR TITLE
W&B wandb support

### PR DESCRIPTION
**Pull request by msaroufim**
This PR adds experimental wandb support, not sure this is "landable" considering y'all uses tensorboard by default. Personally I vastly prefer wandb because I can share my training runs with a link and don't need to muck around with ssh tunneling so I'm just opening this since I'm using it myself. If there's interest I can work to land this.

To use this you just kick of a training as usual with `CONFIG_FILE="./train_configs/llama3_8b.toml" ./run_llama_train.sh
` but also run `wandb login` and paste in your token
Screenshot 2024-11-25 at 12 16 20 PM

Changes in logs will look like

Screenshot 2024-11-25 at 12 27 42 PM

Also only slightly related but llama 3 tokenizer is not available on hf anymore so added instructions for 3.1 and 3.2

<details>
<summary>Click here for detailed logs.</summary>
[rank0]:2024-11-25 11:33:24,320 - root - INFO - Dumping traces at step 1000
[rank0]:2024-11-25 11:33:24,576 - root - INFO - Finished dumping traces in 0.26 seconds
[rank0]:2024-11-25 11:33:24,577 - root - INFO - Sleeping 2 seconds for other ranks to complete
[rank0]:wandb:                                                                                
[rank0]:wandb: 
[rank0]:wandb: Run history:
[rank0]:wandb: loss_metrics/global_avg_loss █▆▅▄▄▃▃▃▃▃▃▃▂▂▂▂▂▂▂▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
[rank0]:wandb: loss_metrics/global_max_loss █▇▄▄▃▃▄▃▃▆▃▃▃▃▃▃▂▂▂▂▃▂▂▃▁▂▂▂▁▃▂▁▂▁▂▂▁▄▁▁
[rank0]:wandb:         memory/max_active(%) ▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
[rank0]:wandb:       memory/max_active(GiB) ▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
[rank0]:wandb:       memory/max_reserved(%) ▁███████████████████████████████████████
[rank0]:wandb:     memory/max_reserved(GiB) ▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
[rank0]:wandb:     memory/num_alloc_retries ▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
[rank0]:wandb:              memory/num_ooms ▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
[rank0]:wandb:                       mfu(%) ▁███████▇██████▇█████████▇█▇████████████
[rank0]:wandb:                         step ▁▁▁▁▂▂▂▂▂▂▃▃▃▃▃▃▃▄▄▄▄▄▅▅▅▆▆▆▆▇▇▇▇▇▇▇████
[rank0]:wandb: time_metrics/data_loading(%) ▁▁▁▁▂▁▁▁▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▂▁▁▁▁▁▁▁▁▂▁▁▂▁▁▁▂
[rank0]:wandb: time_metrics/data_loading(s) ▁▁▁▁▁▁▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▂
[rank0]:wandb:   time_metrics/end_to_end(s) ▁▇▇▇▇█▇▇▇█▇▇▇▇▇▇▇▇▇▇▇▇██▇▇▇█▇▇▇▇▇▇█▇█▇▇▇
[rank0]:wandb:                          wps ███▁████▄█▇▅████████▅▄████▇███▇▄████▇██▇
[rank0]:wandb: 
[rank0]:wandb: Run summary:
[rank0]:wandb: loss_metrics/global_avg_loss 4.53519
[rank0]:wandb: loss_metrics/global_max_loss 4.99517
[rank0]:wandb:         memory/max_active(%) 43.33611
[rank0]:wandb:       memory/max_active(GiB) 41.17145
[rank0]:wandb:       memory/max_reserved(%) 52.19301
[rank0]:wandb:     memory/max_reserved(GiB) 49.58594
[rank0]:wandb:     memory/num_alloc_retries 0
[rank0]:wandb:              memory/num_ooms 0
[rank0]:wandb:                       mfu(%) 30.75216
[rank0]:wandb:                         step 1000
[rank0]:wandb: time_metrics/data_loading(%) 1.01461
[rank0]:wandb: time_metrics/data_loading(s) 0.01583
[rank0]:wandb:   time_metrics/end_to_end(s) 1.55993
[rank0]:wandb:                          wps 5251.52034
[rank0]:wandb: 
[rank0]:wandb: 🚀 View run skilled-glitter-1 at: https://wandb.ai/sahancpal-meta/torchtitan/runs/r1zqr75b
</details>

### CODE REVIEW

File: torchtitan/metrics.py [hunk 1]
```diff
@@ -134,27 +180,47 @@ def build_metric_logger(
...
             rank_str = f"rank_{torch.distributed.get_rank()}"
             log_dir = os.path.join(log_dir, rank_str)
 
-    return MetricLogger(log_dir, tag, enable_tb)
+    if enable_wandb and not WANDB_AVAILABLE:
+        logger.warning(
+            "W&B logging requested but wandb package is not installed. Continuing without W&B logging."
+        )
+        enable_wandb = False
+    elif enable_wandb:
```
Comment (joecummings) [inline torchtitan/metrics.py]: Just a nit: W&B logs it's own init statement, so having one here might be overkill.

**Comment by nenomigami** [comment]
I personally use wandb and have integrated it into my code. However, the `watch` functionality of wandb throws an error when applying FSDP2. Do you have any suggestions or ideas for resolving this issue?

**Review by fegin — COMMENTED** [review]
I think it's a good idea to add support for wandb as it continues to gain popularity. Would it be possible to make wandb and tensorboard mutually exclusive? It doesn't worth to log into two places.

File: torchtitan/metrics.py (hunk 1 above)
Comment (msaroufim) [reply torchtitan/metrics.py]: yeah good point, w&b is actually quite loud with emojis

**Comment by msaroufim** [comment]
@fegin sure thing I can look into that, just to be clear on the gap to merge
1. would w&b be an optional dependency or not?
2. I can make the logging mutually exclusive
3. Remove some of the info messages
4. I can add a test but noticed tensorboard isn't tested either 
5. Any thoughts on the config in the toml name, good enough as is?

**Comment by fegin** [comment]
> would w&b be an optional dependency or not?
I would suggest that it is an optional dependency. While wandb becomes increasingly popular, most users are still using tensorboard. 
> I can make the logging mutually exclusive
Thanks, that would be good.
> Remove some of the info messages
> I can add a test but noticed tensorboard isn't tested either
I think a e2e run command with wandb screenshot would be good eough.
> Any thoughts on the config in the toml name, good enough as is?
Name is okay, but please check `config_manager.py` and add the config into it.
Also, please raise an exception if wandb is enabled but is not available. Implicit fallback is not a good idea.

File: README.md [hunk 2]
```diff
@@ -88,11 +88,11 @@ Once you have confirmed access, you can run the following command to download th
...
 
-# Llama 3 or 3.1 tokenizer.model
-python torchtitan/datasets/download_tokenizer.py --repo_id meta-llama/Meta-Llama-3-8B --tokenizer_path "original" --hf_token=...
+# Llama 3.1 tokenizer.model
+python torchtitan/datasets/download_tokenizer.py --repo_id meta-llama/Meta-Llama-3.1-8B --tokenizer_path "original" --hf_token=...
 
-# Llama 2 tokenizer.model
-python torchtitan/datasets/download_tokenizer.py --repo_id meta-llama/Llama-2-13b-hf --hf_token=...
+# Llama 3.2 tokenizer.model
+python torchtitan/datasets/download_tokenizer.py --repo_id meta-llama/Llama-3.2-3B --hf_token=...
```
Comment (tianyu-l) [inline README.md]: This was indeed for Llama 2, not for Llama 3.2.
I think we can remove Llama 2 files if they are not helpful anymore.

File: torchtitan/metrics.py [hunk 3]
```diff
@@ -98,21 +105,60 @@ def build_device_memory_monitor():
...
+        if self.use_wandb:
+            # Transform metrics to include tag if present
+            wandb_metrics = {}
+            for k, v in metrics.items():
+                tag = k if self.tag is None else f"{self.tag}/{k}"
+                wandb_metrics[tag] = v
+            wandb_metrics["step"] = step
+            wandb.log(wandb_metrics)
+
+    def log_memory_stats(self, memory_stats: DeviceMemStats, step: int):
```
Comment (tianyu-l) [inline torchtitan/metrics.py]: this seems to not be called anywhere?

**Review by tianyu-l — COMMENTED** [review]
Thanks for helping add wandb!

File: README.md (hunk 2 above)
Comment (msaroufim) [reply README.md]: I can send a seperate PR deprecating Llama2

**Comment by msaroufim** [comment]
Made the checks mutually exclusive, if both are set I default to WanDB 

```
➜  torchtitan git:(msaroufim/wandb) ✗ CONFIG_FILE="./train_configs/llama3_8b.toml" ./run_llama_train.sh

+ NGPU=7
+ LOG_RANK=0
+ CONFIG_FILE=./train_configs/llama3_8b.toml
+ overrides=
+ '[' 0 -ne 0 ']'
+ CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6
+ PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True
+ torchrun --nproc_per_node=7 --rdzv_backend c10d --rdzv_endpoint=localhost:0 --local-ranks-filter 0 --role rank --tee 3 train.py --job.config_file ./train_configs/llama3_8b.toml
W1202 12:59:06.793000 2761933 site-packages/torch/distributed/run.py:793] 
W1202 12:59:06.793000 2761933 site-packages/torch/distributed/run.py:793] *****************************************
W1202 12:59:06.793000 2761933 site-packages/torch/distributed/run.py:793] Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
W1202 12:59:06.793000 2761933 site-packages/torch/distributed/run.py:793] *****************************************
[rank0]:Both TensorBoard and WandB logging were enabled. Using WandB only.
[rank0]:2024-12-02 12:59:11,652 - root - INFO - Starting job: Llama 3 8B training
[rank0]:2024-12-02 12:59:11,653 - root - INFO - Deterministic training off
[rank0]:2024-12-02 12:59:12,672 - root - WARNING - ENV[TORCH_NCCL_ASYNC_ERROR_HANDLING] = 1 will be overridden to 3 based on job config
[rank0]:2024-12-02 12:59:12,674 - root - INFO - <built-in method upper of str object at 0x7fd6af8e70b0> capacity: NVIDIA H100 (0) with 95.00GiB memory
[rank0]:2024-12-02 12:59:12,732 - root - INFO - Peak FLOPS used for computing MFU: 9.890e+14
[rank0]:2024-12-02 12:59:12,732 - root - INFO - Building 1-D device mesh with ['dp'], [7]
[rank0]:2024-12-02 12:59:12,733 - root - INFO - Building tiktoken tokenizer locally from ./torchtitan/datasets/tokenizer/original/tokenizer.model
[rank0]:2024-12-02 12:59:12,874 - root - INFO - TikTokenizer built: #words 128256, BOS ID 128000, EOS ID 128001
[rank0]:2024-12-02 12:59:12,875 - root - INFO - Preparing c4 dataset from allenai/c4
[rank0]:2024-12-02 12:59:22,127 - root - INFO - Building llama3 8B with ModelArgs(dim=4096, n_layers=32, n_heads=32, n_kv_heads=8, vocab_size=128256, multiple_of=1024, ffn_dim_multiplier=1.3, norm_eps=1e-05, rope_theta=500000, max_seq_len=8192, depth_init=True, norm_type='rmsnorm')
[rank0]:2024-12-02 12:59:22,265 - root - INFO - Model llama3 8B size: 8,030,261,248 total parameters
[rank0]:2024-12-02 12:59:22,266 - root - INFO - Applied selective activation checkpointing to the model
[rank0]:2024-12-02 12:59:22,330 - root - INFO - Applied FSDP to the model
[rank0]:NCCL version 2.21.5+cuda12.4
[rank0]:/home/marksaroufim/.conda/envs/titan/lib/python3.10/site-packages/torch/nn/init.py:51: UserWarning: No PYTORCH_KERNEL_CACHE_PATH or HOME environment variable set! This disables kernel caching. (Triggered internally at ../aten/src/ATen/native/cuda/jit_utils.cpp:1426.)
[rank0]:  tensor.erfinv_()
[rank0]:2024-12-02 12:59:30,682 - root - INFO - <built-in method upper of str object at 0x7fd6af8e70b0> memory usage for model: 4.30GiB(4.52%)
[rank0]:wandb: Using wandb-core as the SDK backend.  Please refer to https://wandb.me/wandb-core for more information.
[rank0]:wandb: Currently logged in as: formalsystem (sahancpal-meta). Use `wandb login --relogin` to force relogin
[rank0]:wandb: WARNING Path ./outputs/tb/20241202-1259/wandb/ wasn't writable, using system temp directory.
[rank0]:wandb: Tracking run with wandb version 0.18.7
[rank0]:wandb: Run data is saved locally in /tmp/wandb/run-20241202_125931-yjcx6og9
[rank0]:wandb: Run `wandb offline` to turn off syncing.
[rank0]:wandb: Syncing run decent-feather-10
[rank0]:wandb: ⭐️ View project at https://wandb.ai/sahancpal-meta/torchtitan
[rank0]:wandb: 🚀 View run at https://wandb.ai/sahancpal-meta/torchtitan/runs/yjcx6og9
[rank0]:2024-12-02 12:59:31,762 - root - INFO - Training starts at step 1, with local batch size 1, global batch size 7, sequence length 8192, total steps 1000 (warmup 200)
[rank0]:2024-12-02 12:59:31,763 - root - INFO - Profiling active. Traces will be saved at ./outputs/profile_trace
[rank0]:2024-12-02 12:59:39,887 - root - INFO - step:  1  loss: 12.2380  memory: 43.22GiB(45.49%)  wps: 1,008  mfu: 5.91%
[rank0]:2024-12-02 12:59:39,887 - root - INFO - Synchronizing and adjusting timeout for all ProcessGroups to 0:01:40
[rank0]:2024-12-02 12:59:53,798 - root - INFO - step: 10  loss: 10.7523  memory: 51.77GiB(54.50%)  wps: 5,300  mfu: 31.04%
[rank0]:2024-12-02 13:00:09,305 - root - INFO - step: 20  loss:  9.0168  memory: 51.77GiB(54.50%)  wps: 5,285  mfu: 30.95%
```

File: README.md (hunk 2 above)
Comment (tianyu-l) [reply README.md]: Sounds good. Can we revert this change for now, as torchtitan doesn't support Llama 3.2 atm?

File: torchtitan/config_manager.py [hunk 4]
```diff
@@ -625,6 +644,17 @@ def _validate_config(self) -> None:
...
+        if hasattr(self.metrics, "enable_tensorboard") and hasattr(
+            self.metrics, "enable_wandb"
+        ):
+            if self.metrics.enable_tensorboard and self.metrics.enable_wandb:
+                logger.warning(
+                    "Both TensorBoard and WandB logging were enabled. Using WandB only."
+                )
+                # Modify the config to disable TensorBoard
+                self.metrics.enable_tensorboard = False
+
```
Comment (tianyu-l) [inline torchtitan/config_manager.py]: I have slightly different opinion on this. I think we should let user control whether they want to use either or both. In rare cases, they may find enabling both to be useful (e.g. monitoring on wandb, but still keeping the TB logs).

File: torchtitan/metrics.py [hunk 5]
```diff
@@ -16,6 +16,13 @@
 from torchtitan.parallelisms import ParallelDims
 from torchtitan.utils import device_module, device_type
 
+# Optional wandb import
+try:
+    import wandb
+
+    WANDB_AVAILABLE = True
+except ImportError:
+    WANDB_AVAILABLE = False
```
Comment (tianyu-l) [inline torchtitan/metrics.py]: let's raise the exception similar to https://github.com/pytorch/torchtitan/blob/main/torchtitan/float8.py#L46
to avoid silently missing the logs

File: torchtitan/metrics.py [hunk 6]
```diff
@@ -93,26 +100,55 @@ def build_device_memory_monitor():
...
+        self.use_wandb = False
+
+        if enable_wandb and WANDB_AVAILABLE:
+            self.use_wandb = True
+            if wandb.run is None:
+                project_name = (
+                    wandb_config.get("project", "torchtitan")
+                    if wandb_config
+                    else "torchtitan"
+                )
```
Comment (tianyu-l) [inline torchtitan/metrics.py]: Since you've specify `torchtitan` as default project in `config_manager.py`, no need to test if it's empty?

File: torchtitan/metrics.py [hunk 7]
```diff
@@ -93,26 +100,55 @@ def build_device_memory_monitor():
...
-        if self.writer is not None:
+        """Log metrics to the configured backend."""
+        if self.use_wandb:
+            wandb_metrics = {
+                (k if self.tag is None else f"{self.tag}/{k}"): v
+                for k, v in metrics.items()
+            }
+            wandb_metrics["step"] = step
+            wandb.log(wandb_metrics)
+        elif self.writer is not None:
```
Comment (tianyu-l) [inline torchtitan/metrics.py]: for readability, we can change this line to `elif self.enable_tb` and then `assert self.writer is not None`

File: torchtitan/metrics.py [hunk 8]
```diff
@@ -126,35 +162,45 @@ def _get_metrics_rank(parallel_dims: ParallelDims) -> int:
...
-            log_dir = os.path.join(log_dir, rank_str)
+    metrics_config = job_config.metrics
+    log_dir = os.path.join(
+        dump_dir, metrics_config.save_tb_folder, datetime.now().strftime("%Y%m%d-%H%M")
+    )
+
+    enable_tb = metrics_config.enable_tensorboard
+    enable_wandb = metrics_config.enable_wandb
+    wandb_config = (
+        metrics_config.wandb_config if hasattr(metrics_config, "wandb_config") else None
```
Comment (tianyu-l) [inline torchtitan/metrics.py]: ditto: since `config_manager` already has this config, we don't need to test its existence.

File: train_configs/llama3_8b.toml [hunk 9]
```diff
@@ -14,6 +14,8 @@ profile_freq = 100
 log_freq = 10
 enable_tensorboard = true
 save_tb_folder = "tb"
+enable_wandb = true
+wandb_config = { project = "torchtitan", entity = "your-team" }
```
Comment (tianyu-l) [inline train_configs/llama3_8b.toml]: to increase visibility, let's move this to `debug_model.toml` (and default to false)

File: README.md (hunk 2 above)
Comment (msaroufim) [reply README.md]: https://github.com/pytorch/torchtitan/pull/712

**Review by tianyu-l — COMMENTED** [review]
Would you please also add a short note on how to use W&B? similar to 
https://github.com/pytorch/torchtitan/blob/main/README.md?plain=1#L63
and
https://github.com/pytorch/torchtitan/blob/main/README.md?plain=1#L103

It might make sense to have a separate `.md` in `docs/` containing both TB and W&B instructions.

File: torchtitan/config_manager.py (hunk 4 above)
Comment (msaroufim) [reply torchtitan/config_manager.py]: Alright that's easy to fix and it's my preference as well but @fegin for visibility as well

Comment (fegin) [reply torchtitan/config_manager.py]: hmm, I think unless they are showing different information that can be useful, I'm not sure why we need to enable both and for performance, this is going to be bad (but probably not noticeable). But I don't have a strong opinion if people prefer to enable both.

File: torchtitan/metrics.py [hunk 10]
```diff
@@ -126,35 +162,45 @@ def _get_metrics_rank(parallel_dims: ParallelDims) -> int:
...
+        metrics_rank = _get_metrics_rank(parallel_dims)
+        is_metrics_rank = torch.distributed.get_rank() == metrics_rank
+        enable_tb = enable_tb and is_metrics_rank
+        enable_wandb = enable_wandb and is_metrics_rank
+    else:
+        rank_str = f"rank_{torch.distributed.get_rank()}"
+        log_dir = os.path.join(log_dir, rank_str)
 
-    return MetricLogger(log_dir, tag, enable_tb)
+    return MetricLogger(log_dir, tag, enable_tb, enable_wandb, wandb_config)
```
Comment (carmocca) [inline torchtitan/metrics.py]: From an extensibility perspective, it would be nicer to have this pattern:

```python
    if not should_log:
        return DummyLogger()
    if enable_wandb:
        return WandbLogger(wandb_config)
    if enable_tb:
        return TensorBoardLogger(log_dir, tag)
    raise NotImplementedError
```

With

```python

class DummyLogger:
    def log(self, *args, **kwargs):
        pass

    def close(self):
        pass


class TensorBoardLogger:
    def __init__(self, log_dir, tag):
        self.tag = tag
        self.writer = SummaryWriter(log_dir, max_queue=1000)

    def log(self, metrics: Dict[str, Any], step: int):
        for k, v in metrics.items():
            tag = k if self.tag is None else f"{self.tag}/{k}"
            self.writer.add_scalar(tag, v, step)

    def close(self):
        self.writer.close()
```

Since each class has its own separate logic, it's easier for forks to have their custom loggers with reduced conflicts. It also avoids the problem of optional types inside the classes (avoids assertions).

You could then import `wandb` inside the `WandbLogger` class to avoid importing it during startup

**Comment by msaroufim** [comment]
I'm getting a 503 error with downloading C4 now, so will push some untesteted updates and ask for a final review once I confirm things work

File: torchtitan/metrics.py (hunk 10 above)
Comment (msaroufim) [reply torchtitan/metrics.py]: I like the way this code looks but it goes back to the same issue we're discussing above around whether we should have multiple loggers in case people enable both wandb and tensorboard

Right now implicitly enabling both would enable whatever is the topmost condition

Also more curious but are you deliberately not using inheritance here?

Comment (carmocca) [reply torchtitan/metrics.py]: My bad, I did not look at the existing comments extensively.

For that case, you (or a fork) could add a logger class that takes a list of loggers and simply iterates over them.

```python
    loggers = []
    if should_log:
        if enable_wandb:
            loggers.append(WandbLogger(wandb_config))
        if enable_tb:
            loggers.append(TensorBoardLogger(log_dir, tag))
    return Loggers(loggers)
```
This removes the need for `DummyLogger`, iterating over an empty list does nothing.

However, this only works well if the loggers agree to exposing the same interface under the hood, but my opinion is that this repo should stay simple and not think too much about this. From my previous experience, users don't enable two loggers at the same time.

Comment (msaroufim) [reply torchtitan/metrics.py]: I think you've convinced me, 2 loggers seems clunky - incorporated your feedback

File: torchtitan/metrics.py (hunk 5 above)
Comment (msaroufim) [reply torchtitan/metrics.py]: Went for something a bit different here

[rank0]:2024-12-02 16:16:43,791 - root - ERROR - Failed to create WandB logger: No module named 'wandb'. Please install it using 'pip install wandb'.

**Review by carmocca — APPROVED** [review]
The metrics.py structure LGTM now. Thanks!

File: docs/metrics.md [hunk 11]
```diff
@@ -0,0 +1,36 @@
+# Metrics
+
+We support automatically collecting metrics such as
+1. High level system metrics such as MFU, average loss, max loss and words per second along with some
+2. Memory metrics to measure max VRAM consumption and the number of OOMs
+3 Timing metrics to measure data loading bottlenecks
```
Comment (tianyu-l) [inline docs/metrics.md]:
```suggestion
3. Timing metrics to measure data loading bottlenecks
```

File: torchtitan/metrics.py [hunk 12]
```diff
@@ -88,31 +87,68 @@ def reset_peak_stats(self):
 
 
 def build_device_memory_monitor():
-    device_memory_monitor = DeviceMemoryMonitor(device_type)
+    device_memory_monitor = DeviceMemoryMonitor()
```
Comment (tianyu-l) [inline torchtitan/metrics.py]: I think we still need `device_type`, as it is imported on line 17.
But I agree it's a bit confusing, maybe should use different naming.

File: torchtitan/metrics.py [hunk 13]
```diff
@@ -88,31 +87,68 @@ def reset_peak_stats(self):
...
+        f"{device_type.upper()} capacity: {device_memory_monitor.device_name}"
         f"with {device_memory_monitor.device_capacity_gib:.2f}GiB memory"
     )
-
     return device_memory_monitor
 
 
-class MetricLogger:
-    def __init__(self, log_dir, tag, enable_tb):
+class DummyLogger:
```
Comment (tianyu-l) [inline torchtitan/metrics.py]: We can still do inheritance here right? Can call it `BaseLogger` if inheriting from a dummy one sounds confusing.
Btw I'm OK with only one logger if it makes the code simple.

File: torchtitan/metrics.py [hunk 14]
```diff
@@ -126,35 +162,69 @@ def _get_metrics_rank(parallel_dims: ParallelDims) -> int:
...
-    parallelism is enabled, without forcing logging from all ranks to capture loss information.
+    Build an appropriate metric logger based on configuration.
     """
+    metrics_config = job_config.metrics
+
+    # Log initial config state
+    logger.info(
+        f"Building logger with config: wandb={metrics_config.enable_wandb}, "
+        f"tensorboard={metrics_config.enable_tensorboard}"
+    )
```
Comment (tianyu-l) [inline torchtitan/metrics.py]: It seems the `logger.info` from loggers is a bit verbose.
I suggest we keep only two by doing: first being the one here, second inside `__init__` of loggers.
Other loggings, if necessary, can go to `logger.debug` (or `logger.error` for the import error one). What do you think?

File: README.md [hunk 15]
```diff
@@ -99,25 +99,9 @@ Llama 3 8B model locally on 8 GPUs
 CONFIG_FILE="./train_configs/llama3_8b.toml" ./run_llama_train.sh
 ```
 
+## Logging
```
Comment (tianyu-l) [inline README.md]: This would invalidate the hyperlink above from
`Loss, GPU memory, throughput (tokens/sec), and MFU displayed and logged via [TensorBoard](#tensorboard)`

I think we can just move this section to `metrics.md` and remain a link above. E.g.
`Loss, GPU memory, throughput (tokens/sec), and MFU displayed and [logged via TensorBoard or WandB](docs/metrics.md)`

File: torchtitan/metrics.py (hunk 14 above)
Comment (msaroufim) [reply torchtitan/metrics.py]: yeah my b, most of those should be debug logs - that's what I used them for

File: torchtitan/metrics.py [hunk 16]
```diff
@@ -167,14 +167,14 @@ def _get_metrics_rank(parallel_dims: ParallelDims) -> int:
 
 def build_metric_logger(
     job_config: JobConfig, parallel_dims: ParallelDims, tag: Optional[str] = None
-) -> Union[DummyLogger, TensorBoardLogger, WandBLogger]:
+) -> Union[BaseLogger, TensorBoardLogger, WandBLogger]:
```
Comment (tianyu-l) [inline torchtitan/metrics.py]:
```suggestion
) -> BaseLogger:
```

File: torchtitan/metrics.py [hunk 17]
```diff
@@ -90,29 +89,66 @@ def reset_peak_stats(self):
...
-    def log(self, metrics: Dict[str, Any], step: int):
-        if self.writer is not None:
-            for k, v in metrics.items():
-                tag = k if self.tag is None else f"{self.tag}/{k}"
-                self.writer.add_scalar(tag, v, step)
+        self.wandb.init(
+            project="torchtitan",
+            dir=log_dir,
+        )
+        logger.debug("WandB logging enabled")
```
Comment (tianyu-l) [inline torchtitan/metrics.py]:
```suggestion
        logger.info("WandB logging enabled")
```

File: torchtitan/metrics.py [hunk 18]
```diff
@@ -126,35 +162,69 @@ def _get_metrics_rank(parallel_dims: ParallelDims) -> int:
...
+    if metrics_config.rank_0_only and should_log:
+        metrics_rank = _get_metrics_rank(parallel_dims)
+        should_log = torch.distributed.get_rank() == metrics_rank
+
+    logger.debug(
+        f"Logging decision: has_logging_enabled={has_logging_enabled}, should_log={should_log}"
+    )
+
+    if not should_log:
+        logger.debug("Returning DummyLogger due to should_log=False")
```
Comment (tianyu-l) [inline torchtitan/metrics.py]:
```suggestion
        logger.debug("Returning BaseLogger due to should_log=False")
```

**Review by tianyu-l — APPROVED** [review]
Looks awesome! Thank you very much!
Had some final inline comments.

**Comment by msaroufim** [comment]
Thanks again for all the feedback. Had a final lint issue I fixed now. Would you be OK if I merge this or would you rather do so?

File: torchtitan/config_manager.py [hunk 19]
```diff
@@ -121,15 +121,16 @@ def __init__(self):
...
-            help="Whether to enable color printing",
+            default=False,
+            help="Whether to log metrics to TensorBoard",
         )
         self.parser.add_argument(
-            "--metrics.enable_tensorboard",
+            "--metrics.enable_color_printing",
             action="store_true",
-            help="Whether to log metrics to TensorBoard",
+            default=True,
```
Comment (carmocca) [inline torchtitan/config_manager.py:132]: Was flipping the `enable_color_printing` default intended? If so, the action `store_true` does not fit well anymore. It's now true by default
